### PR TITLE
Add resource factory

### PR DESCRIPTION
**Pull request by eliecharra**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | no
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #347
| ❓ Documentation  | no

## Description

* Add ResourceFactory
* Update middlewares to use resource factory and thus resources created from middlewares can now be filtered

### CODE REVIEW

File: pkg/middlewares/aws_instance_block_device_test.go [hunk 1]
```diff
@@ -118,12 +123,21 @@ func TestAwsInstanceBlockDeviceResourceMapper_Execute(t *testing.T) {
 					},
 				},
 			},
+			func(factory *terraform.MockResourceFactory) {
+				factory.On("CreateResource", mock.Anything, "aws_ebs_volume").Times(2).Return(nil, nil)
```
Comment (wbeuil) [inline pkg/middlewares/aws_instance_block_device_test.go:127]: Just FYI, you can use here `Twice()` instead of `Times(2)` but no need to change it.

Comment (eliecharra) [reply pkg/middlewares/aws_instance_block_device_test.go:127]: Yeah I know but I prefer to always use `Times`

Comment (wbeuil) [reply pkg/middlewares/aws_instance_block_device_test.go:127]: If i == 1 you use `Once` and if i > 1 `Times(i)` ?

Comment (eliecharra) [reply pkg/middlewares/aws_instance_block_device_test.go:127]: Yeah and I was wondering myself if I should use `Times` everywhere

**Review by wbeuil — APPROVED** [review]
Nice one ! I will push really soon a PR that start testing those middlewares.

File: pkg/middlewares/aws_instance_block_device_test.go (hunk 1 above)
Comment (wbeuil) [reply pkg/middlewares/aws_instance_block_device_test.go:127]: Indeed, it is more explicit.

**Pull request merged**